linker/descriptors.go

```diff
@@ -1199,10 +1234,22 @@ func (f *fldDescriptor) HasPresence() bool {
 	if f.proto.GetLabel() == descriptorpb.FieldDescriptorProto_LABEL_REPEATED {
 		return false
 	}
-	return f.IsExtension() ||
-		f.Syntax() == protoreflect.Proto2 ||
+	if f.IsExtension() ||
 		f.Kind() == protoreflect.MessageKind || f.Kind() == protoreflect.GroupKind ||
-		f.proto.OneofIndex != nil
+		f.proto.OneofIndex != nil {
+		return true
+	}
+	switch f.Syntax() {
+	case protoreflect.Proto2:
+		return true
+	case protoreflect.Proto3:
+		return false
+	case protoreflect.Editions:
+		fieldPresence := descriptorpb.FeatureSet_FieldPresence(resolveFeature(f, fieldPresenceField).Enum())
```

Comment:
In the Go proto implementation we implemented proto2 and proto3 eventually as editions. This gave us more confidence that the editions implementation was working as expected.

This required surprisingly few changes. It requires two changes. You need to initialize the edition to EDITION_PROTO2/EDITION_PROTO3, load the approrpiate defaults and remove all usages of Syntax() from the code. Infer all semantics from the editions features.
See
https://go-review.googlesource.com/c/protobuf/+/572435
https://go-review.googlesource.com/c/protobuf/+/572695/4

Reply:
To determine repeated field encoding, it's still necessary to look at the field option `packed`, right? Similar for delimited message encoding and legacy required. 

Good to know that the approach is working well for you! We're considering to do the same in our ECMAScript runtime.

Reply:
You don't need the options because `packed` cannot be used with editions. protoc will raise an error if you try to use `[packed = true/false]` on any field.

Reply:
Okay, I did this in cf30e499946f1bb93c410de656cec53b3757c086.

@timostamm, yes, you still have to look at the option first. But it will be unset for editions, so fallback to the feature. I just realized that there was a bunch of superfluous logic in `IsPacked` because I had already made other changes (based on [this comment](https://github.com/bufbuild/protocompile/pull/260#discussion_r1532491673)) but never cleaned up later logic in the proto3 branch. So this commit actually makes this method _way_ simpler.

While searching for other uses of syntax, I caught another one other that needed to be updated to use features. And that one was a bit more complicated to address: we have to know whether an enum is open or closed when interpreting options, in case an enum value is given as an out-of-range number. That check was previously looking at the syntax. It's now been fixed, but the logic isn't as straight-forward because, for editions, we might not yet know if the enum is open or closed, because we are still interpreting options, and the feature that indicates if the enum is open or closed is in an option. So can report an error immediately if the syntax is proto2, but for editions we must defer until we've finished interpreting the options and then check if the enum is open or not.

Reply:
I understand, @lfolger. But a descriptor for a proto2 or proto3 file does not have any edition feature options set. A proto2 field with `[packed = true]` will have FieldOption.packed set to true, but FieldOptions.features.repeated_field_encoding is unset. When treating this descriptor as EDITION_PROTO2, it's still necessary to not only look FieldOption.packed to determine the repeated field encoding of the specific field (or synthesize a FeatureSet).

Reply:
Sorry, I just realized that I completely misunderstood what you meant. Yes this is an issue.

We solved this by overwriting the edition feature when the option was set. This is not possible for you given that you resolve the options later I guess.

Ignore this comment ;)